src/test/java/unit/AiNotebookControlled.java

```diff
@@ -0,0 +1,312 @@
+    TestCommand.runCommandExpectSuccess("workspace", "set", "--id=" + getWorkspaceId());
+
+    // `terra resources create ai-notebook --name=$name`
+    String name = "startStop";
+    TestCommand.runCommandExpectSuccess("resources", "create", "ai-notebook", "--name=" + name);
+    pollDescribeForNotebookState(name, "PROVISIONING");
+
+    // `terra notebooks start --name=$name`
+    // TODO (PF-869): change this to expect success once polling notebook operations is fixed
+    TestCommand.runCommand("notebooks", "start", "--name=" + name);
+    pollDescribeForNotebookState(name, "ACTIVE");
+
+    // `terra notebooks stop --name=$name`
+    // TODO (PF-869): change this to expect success once polling notebook operations is fixed
+    TestCommand.runCommand("notebooks", "stop", "--name=" + name);
+    pollDescribeForNotebookState(name, "STOPPED");
+  }
+
+  /**
+   * Helper method to poll `terra resources describe` until the notebook state equals that
+   * specified. Uses the current workspace.
+   */
+  public static void pollDescribeForNotebookState(String resourceName, String notebookState)
+      throws InterruptedException, JsonProcessingException {
+    pollDescribeForNotebookState(resourceName, notebookState, null);
+  }
+
+  /**
+   * Helper method to poll `terra resources describe` until the notebook state equals that
+   * specified. Filters on the specified workspace id; Uses the current workspace if null.
+   */
+  public static void pollDescribeForNotebookState(
+      String resourceName, String notebookState, UUID workspaceId)
+      throws InterruptedException, JsonProcessingException {
+    HttpUtils.pollWithRetries(
+        () ->
+            workspaceId == null
+                ? TestCommand.runAndParseCommandExpectSuccess(
+                    UFAiNotebook.class, "resources", "describe", "--name=" + resourceName)
+                : TestCommand.runAndParseCommandExpectSuccess(
+                    UFAiNotebook.class,
+                    "resources",
+                    "describe",
+                    "--name=" + resourceName,
+                    "--workspace=" + workspaceId),
+        (result) -> notebookState.equals(result.state),
+        (ex) -> false, // no retries
+        2 * 20, // up to 20 minutes
+        Duration.ofSeconds(30)); // every 30 seconds
+
+    UFAiNotebook describeNotebook =
+        workspaceId == null
+            ? TestCommand.runAndParseCommandExpectSuccess(
+                UFAiNotebook.class, "resources", "describe", "--name=" + resourceName)
+            : TestCommand.runAndParseCommandExpectSuccess(
+                UFAiNotebook.class,
+                "resources",
+                "describe",
+                "--name=" + resourceName,
+                "--workspace=" + workspaceId);
+    assertEquals(notebookState, describeNotebook.state, "notebook state matches");
+    if (!notebookState.equals("PROVISIONING")) {
+      assertNotNull(describeNotebook.proxyUri, "proxy url is populated");
```

Comment:
consider separating the act of polling from a state from asserting what the described notebook should look like. I'm kind of surprised that "STOPPED" has a proxyUri.

Reply:
Agreed they're separate operations. But I think it makes sense to bundle them since they're always called together. I included this here because the create notebooks [tutorial](https://cloud.google.com/notebooks/docs/create-new) says to wait until the "open jupyterlab" link becomes active, which I think is equivalent to polling for the proxy url to be defined. Maybe it only makes sense to check the proxy url when it goes from provisioning -> active, but since it's a kind of status indicator, I thought it made sense to check it when polling for state changes.